File: mobile/android/android-components/components/browser/session/src/main/java/mozilla/components/browser/session/Download.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.browser.session

import android.os.Environment
import mozilla.components.browser.state.state.content.DownloadState
import java.util.UUID

/**
 * Value type that represents a Download.
 *
 * @property url The full url to the content that should be downloaded.
 * @property fileName A canonical filename for this download.
 * @property contentType Content type (MIME type) to indicate the media type of the download.
 * @property contentLength The file size reported by the server.
 * @property userAgent The user agent to be used for the download.
 * @property destinationDirectory The matching destination directory for this type of download.
 * @property referrerUrl The site that linked to this download.
 * @property id a unique ID of this download.
 */
data class Download(
    val url: String,
    val fileName: String? = null,
    val contentType: String? = null,
    val contentLength: Long? = null,
    val userAgent: String? = null,
    val destinationDirectory: String = Environment.DIRECTORY_DOWNLOADS,
    val referrerUrl: String? = null,
    val id: String = UUID.randomUUID().toString()
)

internal fun Download.toDownloadState() = DownloadState(
    url,
    fileName,
    contentType,
    contentLength,
    userAgent,
    destinationDirectory,
    referrerUrl,
    id
)

File: dom/media/test/marionette/yttest/download.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this file,
# You can obtain one at http://mozilla.org/MPL/2.0/.
import sys
from pytube import YouTube


def download_streams(video_id, output_path="data"):
    yt = YouTube("https://youtube.com/watch?v=%s" % video_id)
    for stream in yt.streams.all():
        fn = "%s-%s-%s.%s" % (video_id, stream.itag, stream.type, stream.subtype)
        stream.download(output_path="data", filename=fn)
        print("%s downloaded" % fn)


if __name__ == "__main__":
    download_streams(sys.argv[-1])
